blockprod/src/detail/builder.rs

```diff
@@ -0,0 +1,220 @@
+    Stop,
+    Start,
+}
+
+impl PerpetualBlockBuilder {
+    pub fn new(
+        chain_config: Arc<ChainConfig>,
+        chainstate_handle: ChainstateHandle,
+        mempool_handle: MempoolHandle,
+        time_getter: TimeGetter,
+        builder_rx: mpsc::UnboundedReceiver<BlockBuilderControlCommand>,
+        enabled: bool,
+    ) -> Self {
+        let (block_makers_tx, block_maker_rx) = crossbeam_channel::unbounded();
+        Self {
+            chain_config,
+            chainstate_handle,
+            mempool_handle,
+            time_getter,
+            builder_rx,
+            block_makers_tx: block_makers_tx.clone(),
+            block_maker_rx,
+            enabled,
+            _block_makers_destroyer: BlockMakersDestroyer(block_makers_tx),
+        }
+    }
+
+    pub fn stop_building(
+        &mut self,
+        new_tip_id: Id<Block>,
+        new_tip_height: BlockHeight,
+    ) -> Result<(), BlockProductionError> {
+        self.block_makers_tx.send(BlockMakerControlCommand::StopBecauseNewTip(
+            new_tip_id,
+            new_tip_height,
+        )).expect("The channel can never be disconnected since there's a receiver always alive in self");
+        Ok(())
+    }
+
+    pub fn stop_all_block_makers(&self) -> Result<(), BlockProductionError> {
+        self.block_makers_tx.send(BlockMakerControlCommand::JustStop).expect(
+            "The channel can never be disconnected since there's a receiver always alive in self",
+        );
+        Ok(())
+    }
+
+    pub async fn trigger_new_block_production(
+        &self,
+        current_tip_id: Id<Block>,
+        current_tip_height: BlockHeight,
+    ) -> Result<(), BlockProductionError> {
+        if !self.enabled {
+            return Ok(());
+        }
+
+        let chain_config = self.chain_config.clone();
+        let chainstate_handle = self.chainstate_handle.clone();
+        let mempool_handle = self.mempool_handle.clone();
+        let time_getter = self.time_getter.clone();
+        let command_receiver = self.block_maker_rx.clone();
+        tokio::spawn(async move {
+            BlockMaker::new(
+                chain_config,
+                chainstate_handle,
+                mempool_handle,
+                time_getter,
+                current_tip_id,
+                current_tip_height,
+                command_receiver,
+            )
+            .run()
+            .await
+        });
+        Ok(())
+    }
+
+    pub async fn run(&mut self) -> Result<(), BlockProductionError> {
+        let mut mempool_rx = self.subscribe_to_mempool_events().await?;
+        let mut chainstate_rx = self.subscribe_to_chainstate_events().await?;
+
+        loop {
+            tokio::select! {
+                // when we receive information from chainstate that we have a new tip, we stop building the current block,
+                // and expect the mempool to soon send a trigger for a new command
+                block_info = chainstate_rx.recv().fuse() => {
```

Comment:
What problem does the `fuse` call here solve? I read through the docs and I'm still not quite sure. Is it possible that `poll` "may be called too often" on this future like it says in the documentation? Does this have to do with the `tokio::select!` macro?

Reply:
Yes. Fuse basically makes it possible to poll a future, get the result, and reset again. It "fuses" the Ready state into another Pending state. At least this is how I understand it.